src/XPurchase/VirtoCommerce.XPurchase/Services/CartAvailMethodsService.cs

```diff
@@ -137,7 +138,29 @@ public async Task<IEnumerable<PaymentMethod>> GetAvailablePaymentMethodsAsync(Ca
             }
 
             return result.Results;
-        }
+        }
+
+        public async Task<IEnumerable<LineItem>> GetAvailableGiftsAsync(CartAggregate cartAggr)
+        {
+            var promotionEvalResult = await cartAggr.EvaluatePromotionsAsync();
+
+            var giftRewards = promotionEvalResult.Rewards
+                .OfType<GiftReward>()
+                .Where(reward => reward.IsValid)
+                // .Distinct() is needed as multiplied gifts would be returned otherwise.
+                .Distinct();
+            var productIds = giftRewards.Select(x => x.ProductId).Distinct().ToArray();
+            var productsByIds = (await _cartProductService.GetCartProductsByIdsAsync(cartAggr, productIds)).ToDictionary(x => x.Id);
```

Comment:
Need to extend this logic to return GiftItems for gifts that don't have a product id.